File: src/data/generateData.ts
import type { DesulfurizationData, TimeSeriesPoint, Particle, FurnaceProfile } from './types';

function createFurnaceProfile(): FurnaceProfile {
  const outline: [number, number][] = [
    [-3.0, 6.0],
    [-2.6, 5.5],
    [-2.2, 5.0],
    [-1.9, 4.5],
    [-1.7, 4.0],
    [-1.55, 3.5],
    [-1.45, 3.0],
    [-1.4, 2.5],
    [-1.38, 2.0],
    [-1.37, 1.5],
    [-1.37, 1.0],
    [-1.38, 0.5],
    [-1.4, 0.0],
    [-1.45, -0.5],
    [-1.55, -1.0],
    [-1.7, -1.5],
    [-1.9, -2.0],
    [-2.2, -2.5],
    [-2.6, -3.0],
    [-3.0, -3.5],
    [-3.5, -3.8],
    [-4.0, -4.0],
    [4.0, -4.0],
    [3.5, -3.8],
    [3.0, -3.5],
    [2.6, -3.0],
    [2.2, -2.5],
    [1.9, -2.0],
    [1.7, -1.5],
    [1.55, -1.0],
    [1.45, -0.5],
    [1.4, 0.0],
    [1.38, 0.5],
    [1.37, 1.0],
    [1.37, 1.5],
    [1.38, 2.0],
    [1.4, 2.5],
    [1.45, 3.0],
    [1.55, 3.5],
    [1.7, 4.0],
    [1.9, 4.5],
    [2.2, 5.0],
    [2.6, 5.5],
    [3.0, 6.0],
  ];

  const innerWall: [number, number][] = outline.map(([x, y]) => [
    x * 0.92,
    y * 0.92 + 0.05,
  ]);

  const hotMetalSurface: [number, number][] = [
    [-1.1, 1.2],
    [-0.9, 1.35],
    [-0.5, 1.45],
    [0.0, 1.5],
    [0.5, 1.45],
    [0.9, 1.35],
    [1.1, 1.2],
  ];

  const lancePosition: [number, number] = [0.0, 3.5];
  const lancePath: [number, number][] = [
    [0.0, 6.5],
    [0.0, 5.5],
    [0.0, 4.5],
    [0.0, 3.5],
    [0.0, 2.5],
    [0.0, 1.8],
  ];

  const sensorPositions = [
    { id: 'T-001', x: -1.0, y: 1.0, type: 'temperature' as const },
    { id: 'T-002', x: 0.8, y: 0.5, type: 'temperature' as const },
    { id: 'T-003', x: -0.5, y: -0.5, type: 'temperature' as const },
    { id: 'F-001', x: 0.0, y: 2.0, type: 'flow' as const },
    { id: 'F-002', x: 0.3, y: 1.5, type: 'flow' as const },
    { id: 'L-001', x: -1.2, y: 1.3, type: 'level' as const },
  ];

  return { outline, innerWall, lancePosition, lancePath, hotMetalSurface, sensorPositions };
}

function generateTimeSeries(frameCount: number): TimeSeriesPoint[] {
  const data: TimeSeriesPoint[] = [];
  const S0 = 480;
  const Sf = 28;
  const k1 = 0.004;
  const k2 = 0.0025;

  for (let i = 0; i < frameCount; i++) {
    const t = i;
    const progress = t / frameCount;

    let sulfur_ppm: number;
    let powder_flow: number;
    let argon_flow: number;

    if (progress < 0.08) {
      const phase = progress / 0.08;
      sulfur_ppm = S0 - phase * 5;
      powder_flow = 0;
      argon_flow = 15 + phase * 85;
    } else if (progress < 0.55) {
      const phase = (progress - 0.08) / 0.47;
      sulfur_ppm = (S0 - 5) * Math.exp(-k1 * t * 1.8) + Sf;
      powder_flow = 6.5 + 2.5 * Math.sin(phase * Math.PI) + (Math.random() - 0.5) * 0.8;
      argon_flow = 100 + 30 * Math.sin(phase * Math.PI * 0.7) + (Math.random() - 0.5) * 5;
    } else if (progress < 0.78) {
      const phase = (progress - 0.55) / 0.23;
      sulfur_ppm = Math.max(Sf + 10, (S0 - 5) * Math.exp(-k1 * t * 1.8) * Math.exp(-k2 * (t - frameCount * 0.55) * 0.5));
      powder_flow = 6.5 * (1 - phase * 0.7) + (Math.random() - 0.5) * 0.5;
      argon_flow = 110 + 20 * Math.sin(phase * Math.PI) + (Math.random() - 0.5) * 3;
    } else {
      const phase = (progress - 0.78) / 0.22;
      sulfur_ppm = Sf + 8 * (1 - phase * 0.6) + (Math.random() - 0.5) * 2;
      powder_flow = Math.max(0, 2.0 * (1 - phase));
      argon_flow = 80 * (1 - phase * 0.4) + (Math.random() - 0.5) * 3;
    }

    sulfur_ppm = Math.max(Sf - 5, sulfur_ppm);

    const temperature = 1320 + 40 * Math.sin(progress * Math.PI * 1.5) - progress * 25 + (Math.random() - 0.5) * 3;
    const material_level = 85 - progress * 15 + 5 * Math.sin(progress * Math.PI * 2) + (Math.random() - 0.5) * 2;

    data.push({
      time: t,
      sulfur_ppm: Math.round(sulfur_ppm * 100) / 100,
      sulfur_rate: 0,
      temperature: Math.round(temperature * 10) / 10,
      powder_flow: Math.round(Math.max(0, powder_flow) * 100) / 100,
      argon_flow: Math.round(Math.max(0, argon_flow) * 10) / 10,
      material_level: Math.round(Math.max(0, Math.min(100, material_level)) * 10) / 10,
    });
  }

  for (let i = 1; i < data.length; i++) {
    const dt = data[i].time - data[i - 1].time;
    if (dt > 0) {
      data[i].sulfur_rate = Math.round(((data[i].sulfur_ppm - data[i - 1].sulfur_ppm) / dt) * 100) / 100;
    }
  }

  return data;
}

export function generateParticles(
  timeIndex: number,
  timeSeries: TimeSeriesPoint[],
  furnace: FurnaceProfile,
  particleCount: number = 600
): Particle[] {
  const point = timeSeries[timeIndex];
  if (!point || point.powder_flow <= 0.1) return [];

  const particles: Particle[] = [];
  const intensity = Math.min(1, point.powder_flow / 9);
  const activeCount = Math.floor(particleCount * intensity);

  const [lx, ly] = furnace.lancePosition;
  const spreadX = 1.2 * intensity;
  const spreadY = 1.5 * intensity;

  for (let i = 0; i < activeCount; i++) {
    const seed = i * 7.31 + timeIndex * 0.17;
    const rand1 = Math.abs(Math.sin(seed)) ;
    const rand2 = Math.abs(Math.sin(seed * 1.73 + 0.5));
    const rand3 = Math.abs(Math.sin(seed * 2.37 + 1.2));
    const rand4 = Math.abs(Math.sin(seed * 3.14 + 2.1));

    const angle = (rand1 - 0.5) * Math.PI * 0.8;
    const dist = rand2 * spreadY * (0.3 + 0.7 * rand3);

    const x = lx + Math.sin(angle) * dist * spreadX;
    const y = ly - dist + Math.sin(timeIndex * 0.05 + i * 0.3) * 0.1;

    const isInIron = y < 1.5 && Math.abs(x) < 1.2;
    const concentration = isInIron ? (0.6 + 0.4 * (1 - dist / spreadY)) * intensity : 0.3 * intensity;

    particles.push({
      x,
      y: Math.max(-2.5, y),
      concentration: Math.round(concentration * 100) / 100,
      velocity: Math.round((0.5 + rand4 * 2.0) * intensity * 100) / 100,
      age: rand3,
      size: (2 + rand4 * 4) * (0.5 + concentration * 0.5),
    });
  }

  return particles;
}

export function generateDesulfurizationData(): DesulfurizationData {
  const frameCount = 900;
  const furnace = createFurnaceProfile();
  const timeSeries = generateTimeSeries(frameCount);

  return {
    timeSeries,
    totalDuration: frameCount,
    frameCount,
    furnace,
  };
}

File: scripts/generate_data.py
"""
高炉铁水脱硫数据预处理脚本
使用 Pandas + NumPy 处理现场传感器CSV数据
清洗时间戳、对齐采样率、计算微积分、输出前端JSON
"""

import pandas as pd
import numpy as np
import json
import os
from pathlib import Path


def clean_timestamps(df: pd.DataFrame, ts_col: str = 'timestamp') -> pd.DataFrame:
    df[ts_col] = pd.to_datetime(df[ts_col], errors='coerce')
    df = df.dropna(subset=[ts_col])
    df = df.sort_values(ts_col).reset_index(drop=True)
    return df


def remove_outliers(df: pd.DataFrame, columns: list[str], n_std: float = 4.0) -> pd.DataFrame:
    for col in columns:
        if col in df.columns:
            mean = df[col].mean()
            std = df[col].std()
            lower = mean - n_std * std
            upper = mean + n_std * std
            df.loc[(df[col] < lower) | (df[col] > upper), col] = np.nan
    return df


def align_sampling_rates(dfs: dict[str, pd.DataFrame], target_freq: str = '1S') -> pd.DataFrame:
    resampled = {}
    for name, df in dfs.items():
        df = df.set_index('timestamp')
        df_resampled = df.resample(target_freq).mean()
        df_resampled = df_resampled.interpolate(method='linear', limit=5)
        df_resampled = df_resampled.fillna(method='ffill', limit=3)
        resampled[name] = df_resampled

    aligned = pd.concat(resampled.values(), axis=1)
    aligned = aligned.interpolate(method='linear', limit=5)
    aligned = aligned.fillna(method='ffill', limit=3)
    aligned = aligned.dropna()
    aligned = aligned.reset_index()
    return aligned


def compute_sulfur_kinetics(sulfur_ppm: np.ndarray, dt: float = 1.0) -> tuple[np.ndarray, np.ndarray]:
    sulfur_rate = np.gradient(sulfur_ppm, dt)
    cumulative_desulfurization = np.cumsum(np.abs(sulfur_rate) * dt)
    return sulfur_rate, cumulative_desulfurization


def generate_particle_coordinates(
    lance_x: float, lance_y: float,
    powder_flow: float, argon_flow: float,
    n_particles: int = 600,
    seed: int = 42
) -> list[dict]:
    if powder_flow < 0.1:
        return []

    rng = np.random.default_rng(seed)
    intensity = min(1.0, powder_flow / 9.0)
    active_count = int(n_particles * intensity)

    spread_x = 1.2 * intensity
    spread_y = 1.5 * intensity

    angles = rng.uniform(-np.pi * 0.4, np.pi * 0.4, active_count)
    distances = rng.uniform(0, 1, active_count) * spread_y

    x = lance_x + np.sin(angles) * distances * spread_x
    y = lance_y - distances

    concentrations = np.where(
        (y < 1.5) & (np.abs(x) < 1.2),
        (0.6 + 0.4 * (1 - distances / spread_y)) * intensity,
        0.3 * intensity
    )

    particles = []
    for i in range(active_count):
        particles.append({
            'x': round(float(x[i]), 4),
            'y': round(float(max(-2.5, y[i])), 4),
            'concentration': round(float(concentrations[i]), 4),
            'velocity': round(float(rng.uniform(0.5, 2.5) * intensity), 4),
        })

    return particles


def process_pipeline(data_dir: str, output_dir: str):
    print("Step 1: Reading CSV files...")
    dfs = {}
    csv_files = {
        'temperature': 'temperature_sensors.csv',
        'powder_flow': 'powder_flow_sensors.csv',
        'argon_flow': 'argon_flow_sensors.csv',
        'material_level': 'material_level_sensors.csv',
        'sulfur': 'sulfur_analysis.csv',
    }

    for name, filename in csv_files.items():
        filepath = os.path.join(data_dir, filename)
        if os.path.exists(filepath):
            chunks = []
            for chunk in pd.read_csv(filepath, chunksize=100000):
                chunk = clean_timestamps(chunk)
                chunks.append(chunk)
            dfs[name] = pd.concat(chunks, ignore_index=True)
            print(f"  Loaded {name}: {len(dfs[name])} records")

    if not dfs:
        print("No CSV files found. Generating sample data instead.")
        return generate_sample_data(output_dir)

    print("Step 2: Removing outliers...")
    for name, df in dfs.items():
        numeric_cols = df.select_dtypes(include=[np.number]).columns.tolist()
        dfs[name] = remove_outliers(df, numeric_cols)

    print("Step 3: Aligning sampling rates to 1Hz...")
    aligned = align_sampling_rates(dfs, target_freq='1S')

    print("Step 4: Computing sulfur kinetics...")
    if 'sulfur_ppm' in aligned.columns:
        sulfur_rate, cumul = compute_sulfur_kinetics(aligned['sulfur_ppm'].values)
        aligned['sulfur_rate'] = sulfur_rate
        aligned['cumulative_desulfurization'] = cumul

    print("Step 5: Generating output JSON...")
    output_path = Path(output_dir)
    output_path.mkdir(parents=True, exist_ok=True)

    records = aligned.to_dict(orient='records')
    with open(output_path / 'timeseries.json', 'w', encoding='utf-8') as f:
        json.dump(records, f, ensure_ascii=False)

    print(f"Done. Output {len(records)} records to {output_path / 'timeseries.json'}")
    return aligned


def generate_sample_data(output_dir: str):
    output_path = Path(output_dir)
    output_path.mkdir(parents=True, exist_ok=True)

    n_frames = 900
    S0, Sf = 480, 28
    k1, k2 = 0.004, 0.0025

    timestamps = pd.date_range('2024-06-18 08:00:00', periods=n_frames, freq='1S')

    sulfur_ppm = np.zeros(n_frames)
    powder_flow = np.zeros(n_frames)
    argon_flow = np.zeros(n_frames)

    for i in range(n_frames):
        progress = i / n_frames
        if progress < 0.08:
            phase = progress / 0.08
            sulfur_ppm[i] = S0 - phase * 5
            powder_flow[i] = 0
            argon_flow[i] = 15 + phase * 85
        elif progress < 0.55:
            sulfur_ppm[i] = (S0 - 5) * np.exp(-k1 * i * 1.8) + Sf
            phase = (progress - 0.08) / 0.47
            powder_flow[i] = 6.5 + 2.5 * np.sin(phase * np.pi)
            argon_flow[i] = 100 + 30 * np.sin(phase * np.pi * 0.7)
        elif progress < 0.78:
            sulfur_ppm[i] = max(Sf + 10,
                (S0 - 5) * np.exp(-k1 * i * 1.8) * np.exp(-k2 * (i - n_frames * 0.55) * 0.5))
            phase = (progress - 0.55) / 0.23
            powder_flow[i] = 6.5 * (1 - phase * 0.7)
            argon_flow[i] = 110 + 20 * np.sin(phase * np.pi)
        else:
            phase = (progress - 0.78) / 0.22
            sulfur_ppm[i] = Sf + 8 * (1 - phase * 0.6)
            powder_flow[i] = max(0, 2.0 * (1 - phase))
            argon_flow[i] = 80 * (1 - phase * 0.4)

    sulfur_rate, cumul = compute_sulfur_kinetics(sulfur_ppm)
    temperature = 1320 + 40 * np.sin(np.linspace(0, np.pi * 1.5, n_frames)) - \
                  np.linspace(0, 25, n_frames)
    material_level = 85 - np.linspace(0, 15, n_frames) + \
                     5 * np.sin(np.linspace(0, np.pi * 2, n_frames))

    df = pd.DataFrame({
        'timestamp': timestamps,
        'sulfur_ppm': np.round(sulfur_ppm, 2),
        'sulfur_rate': np.round(sulfur_rate, 2),
        'cumulative_desulfurization': np.round(cumul, 2),
        'temperature': np.round(temperature, 1),
        'powder_flow': np.round(np.maximum(0, powder_flow), 2),
        'argon_flow': np.round(np.maximum(0, argon_flow), 1),
        'material_level': np.round(np.clip(material_level, 0, 100), 1),
    })

    records = df.to_dict(orient='records')
    with open(output_path / 'sample_timeseries.json', 'w', encoding='utf-8') as f:
        json.dump(records, f, ensure_ascii=False, default=str)

    print(f"Sample data generated: {len(records)} records")
    return df


if __name__ == '__main__':
    import sys
    data_dir = sys.argv[1] if len(sys.argv) > 1 else './data/csv'
    output_dir = sys.argv[2] if len(sys.argv) > 2 else './public/data'
    process_pipeline(data_dir, output_dir)
